File: 2.Strings/C/reverseString.c
/**
 * Reverse String Program
 * 
 * This program demonstrates how to reverse a string by printing its characters
 * in reverse order. It uses a character array and iterates from the end to the beginning.
 * 
 * Algorithm:
 * 1. Define a character array with the input string
 * 2. Iterate through the array from the last character to the first
 * 3. Print each character during iteration
 * 
 * Time Complexity: O(n) where n is the length of the string
 * Space Complexity: O(1) as we only use the original array
 * 
 * Input: A hardcoded character array - {'s','a','m','i','r'}
 * Output: The reversed string - "rimas"
 */

#include <stdio.h>
#include <string.h>  // For strlen function

/**
 * reverseAndPrint - Reverses and prints a character array
 * @str: The character array to reverse and print
 * @length: The length of the character array
 */
void reverseAndPrint(char str[], int length) {
    printf("Original string: ");
    for (int i = 0; i < length; i++) {
        printf("%c", str[i]);
    }
    printf("\n");
    
    printf("Reversed string: ");
    // Iterate from the end to the beginning
    for (int i = length - 1; i >= 0; i--) {
        printf("%c", str[i]);
    }
    printf("\n");
}

int main() {
    // Define the input character array
    char myName[] = {'s', 'a', 'm', 'i', 'r'};
    int length = sizeof(myName) / sizeof(myName[0]);
    
    // Call the function to reverse and print the string
    reverseAndPrint(myName, length);
    
    // Alternative approach with a null-terminated string
    char myNameStr[] = "samir";
    printf("\nAlternative approach with null-terminated string:\n");
    printf("Original string: %s\n", myNameStr);
    
    printf("Reversed string: ");
    for (int i = strlen(myNameStr) - 1; i >= 0; i--) {
        printf("%c", myNameStr[i]);
    }
    printf("\n");
    
    return 0;
}

File: 2.Strings/java/ReverseString.java
/**
 * ReverseString
 * 
 * This class provides methods to reverse a string.
 * Two implementations are provided:
 * 1. Using Java's StringBuilder (reverseString)
 * 2. Manual implementation using two-pointer technique (reverseStringManual)
 * 
 * Time Complexity: O(n) for both methods, where n is the length of the string
 * Space Complexity: 
 * - StringBuilder: O(n) for creating a new StringBuilder and the resulting string
 * - Manual: O(n) for creating the character array and the resulting string
 * 
 * Input: A string
 * Output: The reversed string
 */
public class ReverseString {
    /**
     * Reverses a string using Java's StringBuilder class.
     * 
     * Algorithm:
     * 1. Create a new StringBuilder with the input string
     * 2. Use the built-in reverse() method
     * 3. Convert back to a String
     * 
     * @param str The input string to reverse
     * @return The reversed string
     */
    public static String reverseString(String str) {
        return new StringBuilder(str).reverse().toString();
    }
    
    /**
     * Reverses a string manually using the two-pointer technique.
     * 
     * Algorithm:
     * 1. Convert the string to a character array
     * 2. Initialize two pointers: left at the beginning and right at the end
     * 3. Swap characters at left and right positions
     * 4. Move left pointer forward and right pointer backward
     * 5. Repeat until pointers meet in the middle
     * 6. Create a new string from the modified character array
     * 
     * @param str The input string to reverse
     * @return The reversed string
     */
    public static String reverseStringManual(String str) {
        // Convert string to character array for in-place swapping
        char[] chars = str.toCharArray();
        int left = 0, right = chars.length - 1;
        
        // Swap characters from both ends moving inward
        while (left < right) {
            // Swap characters at left and right positions
            char temp = chars[left];
            chars[left] = chars[right];
            chars[right] = temp;
            
            // Move pointers toward the center
            left++;
            right--;
        }
        
        // Create a new string from the modified character array
        return new String(chars);
    }
    
    /**
     * Main method to test both string reversal implementations with examples.
     * 
     * @param args Command line arguments (not used)
     */
    public static void main(String[] args) {
        // Test strings
        String str1 = "Hello, World!";
        String str2 = "Programming";
        
        // Display original strings
        System.out.println("Original strings:");
        System.out.println(str1);
        System.out.println(str2);
        
        // Test StringBuilder implementation
        System.out.println("\nAfter reversal (using StringBuilder):");
        System.out.println(reverseString(str1));
        
        // Test manual implementation
        System.out.println("\nAfter reversal (manual implementation):");
        System.out.println(reverseStringManual(str2));
    }
}